### PR TITLE
1.0.0-alpha release notes

### PR DESCRIPTION
**Pull request by brson**
A whole lot happened this cycle. I tried to highlight the best stuff. Please review and note important stuff I'm missing or foolish mistakes.

### CODE REVIEW

File: RELEASES.md [hunk 1]
```diff
@@ -1,3 +1,166 @@
...
+      incurred too much code generation for little benefit.
+    * [Struct variants][structvars] are no longer feature-gated.
+    * Trait bounds can be [polymorphic over lifetimes][hrtb]. Also
+      known as 'higher-ranked trait bounds', this crucially allows
+      unboxed closures to work.
+    * Macros invocations surrounded by parens or square brackets and
+      not terminated by a semicolon are [parsed as
+      expressions][macros], which makes expressions like `vec![1i32,
+      2, 3].len()` work as expected.
+    * Trait objects now implement their traits automatically.
```
Comment (sfackler) [inline RELEASES.md]: Might want to mention trait object safety here as well.

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,166 @@
...
+    * Automatically deriving traits is now done with `#[derive(...)]`
+      not `#[deriving(...)]` for [consistency with other naming
+      conventions][derive].
+    * Importing the containing module at the same time as items it
```
Comment (sfackler) [inline RELEASES.md]: "module or enum"

File: RELEASES.md [hunk 3]
```diff
@@ -1,3 +1,166 @@
...
+      [conventions][coll3] for collections types has resulted in API
+      improvements throughout the standard library.
+    * New [APIs for error handling][err] provide ergonomic interop
+      between error types, and [new conventions][err-conv] describe
+      more clearly the recommended error handling strategies in Rust.
+    * The `fail!` macro has been renamed to [`panic!`][panic] so that
+      it is easier to discuss failure in the context of error handling
+      without making clarifications as to whether you are referring to
+      the 'fail' macro or failure more generally.
+    * On Linux, `OsRng` prefers the new, more reliable `getrandom'
```
Comment (sfackler) [inline RELEASES.md]: "'" should be a "`"

**Comment by jroesch** [comment]
It might be interesting to reference progress on experimental/feature-gated parts of the language for those who don't follow the GitHub and IRC chatter closely. For example a ton of work was done on associated types since 0.12.0 and without closely following the community I would have no idea that they are much much more stable then they were and ready to be experimented with and hammered on.

File: RELEASES.md [hunk 4]
```diff
@@ -1,3 +1,166 @@
...
+    * Closures have been [completely redesigned][unboxed] to be
+      implemented in terms of traits, can now be used as generic type
+      bounds and thus monomorphized and inlined, or via an opaque
+      pointer (boxed) as in the old system. The new system is often
+      referred to as 'unboxed' closures.
+    * Enum variants are [namespaced by their type names][enum].
+    * [`where` clauses][where] provide a more versatile and attractive
+      syntax for specifying generic bounds, though the previous syntax
+      remains valid.
+    * Rust again picks a [fallback] (either i32 or f64) for uninferred
```
Comment (sanxiyn) [inline RELEASES.md]: Link to fallback RFC doesn't parse as intended.

File: RELEASES.md [hunk 5]
```diff
@@ -1,3 +1,166 @@
...
+      numeric types.
+    * Rust [no longer has a runtime][rt] of any description, and only
+      supports OS threads, not green threads.
+    * At long last, Rust has been overhauled for 'dynamically-sized
+      types' ([DST]), which integrates 'fat pointers' (object types,
+      arrays, and `str`) more deeply into the type system, making it
+      more consistent.
+    * Rust now has a general [range syntax][range], `i..j`, `i..`, and
+      `..j` that produce range types and which, when combined with the
+      `Index` operator and multitispatch, leads to a convenient slice
```
Comment (parir) [inline RELEASES.md]: s/multitispatch/multidispatch

File: RELEASES.md (hunk 4 above)
Comment (killercup) [inline RELEASES.md]: Should be `[fallback][fallback]` so the `(either i32 or f64)` part isn't parsed as the link's URL.

**Comment by brson** [comment]
Current feedback has been addressed.

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,174 @@
...
+
+  * ~2400 changes, numerous bugfixes
+
+  * Highlights
+
+    * The language itself is considered feature complete for 1.0,
+      though there is a significant amount of cleanup and bugfixes
+      remaining.
+    * Nearly 50% of the public API surface of the standard library has
+      been declared 'stable'. Those interfaces will not change.
```
Comment (huonw) [inline RELEASES.md]: Should "will not" this be softened slightly? It seems to me we still want some room for tweaks and adjustments based on feedback.

File: RELEASES.md [hunk 7]
```diff
@@ -1,3 +1,180 @@
...
+      N]`][arrays].
+    * The `Copy` trait is no longer implemented automatically. Unsafe
+      pointers no longer implement `Sync` and `Send` so types
+      containing them don't automatically either. `Sync` and `Send`
+      are now 'unsafe traits' so one can "forcibly" implement them via
+      `unsafe impl` if a type confirms to the requirements for them
+      even though the internals do not (e.g. structs containing unsafe
+      pointers like `Arc`). These changes are intended to prevent some
+      footguns and are collectively known as [opt-in built-in
+      traits][oibit] (though `Sync` and `Share` will soon become pure
```
Comment (pythonesque) [inline RELEASES.md]: Shouldn't this be `Sync` and `Send`?

Comment (brson) [inline RELEASES.md]: Good catch.

**Pull request merged**